lib/screens/config/cache.ex

```diff
@@ -0,0 +1,180 @@
+defmodule Screens.Config.Cache do
+  @moduledoc """
+  Functions to read data from a cached copy of the screens config.
+  """
+
+  alias Screens.Config
+  alias ScreensConfig.Devops
+
+  use Screens.Cache.Client, table: :screens_config
+
+  @type table_contents :: list(table_entry)
+
+  @type table_entry ::
+          {{:screen, screen_id :: String.t()}, ScreensConfig.Screen.t()}
+          | {:last_deploy_timestamp, DateTime.t()}
+          | {:devops, ScreensConfig.Devops.t()}
+
+  def ok?, do: table_exists?()
+
+  def refresh_if_loaded_before(screen_id) do
+    with_table default: nil do
+      [[last_deploy_timestamp]] = :ets.match(@table, {:last_deploy_timestamp, :"$1"})
+
+      refresh_if_loaded_before =
+        case :ets.match(@table, {screen_id, %{refresh_if_loaded_before: :"$1"}}) do
+          [[timestamp]] -> timestamp
+          [] -> nil
+        end
+
+      case {last_deploy_timestamp, refresh_if_loaded_before} do
+        {nil, nil} ->
+          nil
+
+        {nil, refresh_if_loaded_before} ->
+          refresh_if_loaded_before
+
+        {last_deploy_timestamp, nil} ->
+          last_deploy_timestamp
+
+        {last_deploy_timestamp, refresh_if_loaded_before} ->
+          Enum.max([last_deploy_timestamp, refresh_if_loaded_before], DateTime)
+      end
+    end
+  end
+
+  def last_deploy_timestamp do
+    with_table do
+      [[last_deploy_timestamp]] = :ets.match(@table, {:last_deploy_timestamp, :"$1"})
+      last_deploy_timestamp
+    end
+  end
+
+  def service_level(screen_id) do
+    with_table default: 1 do
+      case :ets.match(@table, {{:screen, screen_id}, %{app_params: %{service_level: :"$1"}}}) do
+        [[service_level]] -> service_level
+        [] -> 1
+      end
+    end
+  end
+
+  def disabled?(screen_id) do
+    with_table default: false do
+      case :ets.match(@table, {{:screen, screen_id}, %{disabled: :"$1"}}) do
+        [[disabled]] -> disabled
+        [] -> false
+      end
+    end
+  end
+
+  def screen(screen_id) do
+    with_table default: nil do
+      case :ets.match(@table, {{:screen, screen_id}, :"$1"}) do
+        [[screen]] -> screen
+        [] -> nil
+      end
+    end
+  end
+
+  def app_params(screen_id) do
+    with_table default: nil do
+      case :ets.match(@table, {{:screen, screen_id}, %{app_params: :"$1"}}) do
+        [[app_params]] -> app_params
+        [] -> nil
+      end
+    end
+  end
+
+  def devops do
+    with_table default: nil do
+      case :ets.match(@table, {:devops, :"$1"}) do
+        [[devops]] -> devops
+        [] -> nil
+      end
+    end
+  end
+
+  @doc """
+  Returns a list of all screen IDs.
+  """
+  def screen_ids do
+    with_table default: [] do
+      # Not sure how to avoid matching against the struct this way.
```

Comment:
Oh my bad, I fixed this just before posting for review and forgot to remove this comment.

Table entries containing screen configs now look like `{:screen, screen_id}, screen_config}` so they're easy to find.
Previously it was just `{screen_id, screen_config}` so I had to find them by checking if the key was a string, or if the value was a `%Screen{}` struct.

I'll delete this comment.

Reply:
I was wary of using `fun2ms` since it seems to be geared toward erlang code specifically. It uses erlang's version of macros ("parse transforms"). I'm not sure if it would work correctly with elixir code.